samples: cmsis_rtos_v2: Add sample application

spoorthik opened the pull request:
This PR includes sample applications to demonstrate CMSIS RTOS V2 APIs usage.

rgundi reviewed (commented):
LGTM. Please fix the documentation and shippable failures.

dbkinder commented on doc/porting/cmsis_rtos_v2.rst:
```diff
@@ -0,0 +1,85 @@
+generic tool interfaces. Though it was originally defined for ARM Cortex-M
+microcontrollers alone, it could be easily extended to other microcontrollers
+making it generic. For more information on CMSIS RTOS v2, please refer
+http://www.keil.com/pack/doc/CMSIS/RTOS2/html/index.html
```
```suggestion
to the `CMSIS-RTOS2 Documentation <http://www.keil.com/pack/doc/CMSIS/RTOS2/html/index.html>`_.
```

spoorthik replied:
@dbkinder: This will be addressed in @rgundi's PR (#10954).

dbkinder commented on doc/porting/cmsis_rtos_v2.rst:
```diff
@@ -0,0 +1,85 @@
+
+Kernel
+   osKernelGetState, osKernelSuspend, osKernelResume are not supported.
+   Also, osKernelInitialize and osKernelStart are pretty much empty.
```
I suspect these two APIs should just be added to the "not supported" list in the previous sentence. What does "pretty much empty" mean?

spoorthik replied:
@rgundi: As this change is part of your commits, would you please comment on?

rgundi replied:
sure. will address it in my PR.

dbkinder commented on doc/porting/cmsis_rtos_v2.rst:
```diff
@@ -0,0 +1,85 @@
+
+Mutex
+   osMutexPrioInherit is supported by default and is not configurable,
+   we cannot select/unselect this attribute.
```
```suggestion
   you cannot select/unselect this attribute.
```

spoorthik replied:
@dbkinder: This will be addressed in @rgundi's PR (#10954).

dbkinder commented on samples/portability/cmsis_rtos_v2/README.rst:
```diff
@@ -0,0 +1,56 @@
+.. _cmsis_rtos_v2-sample:
+
+Dining Philosophers
```
For the table of contents, since there's already a sample called "Dining Philosophers" change this doc's title to something like "Dining Philosophers (CMSI RTOS V2 APIs)"

spoorthik replied:
@dbkinder: Done, thanks.

dbkinder commented on samples/portability/cmsis_rtos_v2/README.rst:
```diff
@@ -0,0 +1,56 @@
+Overview
+********
+
+An implementation of a solution to the Dining Philosophers problem (a classic
```
Use a complete sentence, and lets add a reference to the "problem":
```suggestion
This sample implements a solution to the `Dining Philosophers problem
<https://en.wikipedia.org/wiki/Dining_philosophers_problem>`_ (a classic
```

dbkinder commented on samples/samples.rst:
```diff
@@ -20,6 +20,7 @@ Samples and Demos
    application_development/*
    display/*
    shields/*
+   portability/*
```
For consistency with the other sample sections (such as display and shields), you'll also need to create a file ``portability/portability.rst`` with this content (resolving the doc build errors you're seeing from shippable):
```
.. _portability-samples:

Portability Samples
###################

.. toctree::
   :maxdepth: 1
   :glob:

   **/*

```

spoorthik replied:
@dbkinder: 👍 thanks.

spoorthik commented:
@dbkinder: Addressed your review comments related to samples. And rest will be addressed in @rgundi's PR.

@rgundi: Shippable is passing now. 

Please review.

rgundi reviewed (approved):
LGTM

dbkinder reviewed (approved):
doc changes LGTM, thanks.

spoorthik commented:
@nashif : Can you please review?

spoorthik commented:
recheck

spoorthik commented:
recheck

spoorthik committed:
samples: cmsis_rtos_v2: Demo with Dining Philosopher's algorithm

The sample application to demonstrate dining philosopher's problem
implementation using CMSIS RTOS V2 APIs with semaphores and mutexes
as resources (forks).

The application makes use of CMSIS_RTOS_V2 APIs on threads, semaphores
and mutexes.

Signed-off-by: Spoorthi K <spoorthi.k@intel.com>

spoorthik committed:
samples: cmsis_rtos_v2: Demo on msgq and timers synchronization

The sample application demonstrates usage of threads, message queue
and timer APIs in CMSIS RTOS V2.

Signed-off-by: Spoorthi K <spoorthi.k@intel.com>
